Added the MNIST dataset for Swift for TensorFlow Datasets Library

pruthiviraj71 opened the pull request:
Suggest any changes if needed.

rxwei commented on Sources/DeepLearning/MNISTdataset.swift:
```diff
@@ -0,0 +1,91 @@
+let testCount = 10000
+
+// Function to download and extract from online
+func download_and_extract(){
```
To ensure consistency and readability, please use camel case for all names in Swift, including function names and variable names.

All our library code is following the [Google Swift Style](https://google.github.io/swift) with the exception that indentation is 4 spaces long. It’d be great if you take a look!

saeta commented on Sources/DeepLearning/MNISTdataset.swift:
```diff
@@ -0,0 +1,91 @@
+}
+
+download_and_extract()
+let (trainx, trainy, testx, testy) = readMNIST(trainImages: trainImages, trainLabels: trainLabels, testImages: testImages, testLabels: testLabels)
```
Newline at the end of the file.

saeta replied:
I think you'll likely want to remove these last couple lines from the file, in order to avoid downloading and extracting the MNIST dataset whenever anyone uses Swift for TensorFlow. Does that make sense?

saeta commented on Sources/DeepLearning/MNISTdataset.swift:
```diff
@@ -0,0 +1,91 @@
+    let trainrowCount = Int32(trainLabels.count)
+    let testrowCount = Int32(testLabels.count)
+    let columnCount = Int32(testImages.count) / testrowCount
+    print(trainImages.count)
```
Could you remove debugging prints?

saeta commented on Sources/DeepLearning/MNISTdataset.swift:
```diff
@@ -0,0 +1,91 @@
+import Python
+import TensorFlow
+
+let np = Python.import("numpy")
```
Do you think you could do a native-swift implementation, instead of falling back to Python?

pruthiviraj71 replied:
I am not sure if there is any native function in Foundation library to decompress gz files which is the file hosted in MNIST official URL. I can do a pure swift implementation for opening the files and converting the file into swift array. Just that I may have to use a third party framework for extraction process.

saeta commented on Sources/DeepLearning/MNISTdataset.swift:
```diff
@@ -0,0 +1,91 @@
+            f_in = gzip.open(helperFiles[counter], "rb")
+            f_out = Python.open(extractedFiles[counter], "wb")
+            sh.copyfileobj(f_in, f_out)
+            print("EXtraction Completed: \(extractedFile)")
```
Looks like you have a typo here.

saeta commented on Sources/DeepLearning/MNISTdataset.swift:
```diff
@@ -0,0 +1,91 @@
+
+// Reads a file into an array of bytes.
+func readFile(_ filename: String) -> [UInt8] {
+    let d = Python.open(filename, "rb").read()
```
Please use more descriptive variable names. :-)

saeta commented on Sources/DeepLearning/MNISTdataset.swift:
```diff
@@ -0,0 +1,91 @@
+let gzip = Python.import("gzip")
+let sh = Python.import("shutil")
+
+let URL = "http://yann.lecun.com/exdb/mnist/"
```
To avoid polluting the namespace, could you define these constants within an object?

pruthiviraj71 replied:
Hey @saeta , I have made all changes as you requested.

pruthiviraj71 commented:
1. Script is now placed under Scripts Directory in root as suggested by @rxwei .
2. Using camel case for all variable and function names as per the [Google Swift Style guide](https://google.github.io/swift/).
3. Removed debugging printing statements and fixed few typos.
4. Added a dictionary MNIST constants to store constants to prevent namespace polluting as suggested by @saeta .

marcrasi commented on Scripts/MNISTdataset.swift:
```diff
@@ -0,0 +1,97 @@
+/*
```
I suggest putting the usage comments on the function declarations instead of at the top level of the file (https://google.github.io/swift/#file-comments).

Also, I have a few more comment formatting requests from https://google.github.io/swift/#documentation-comments:
* Use "///" for documentation comments instead of "/* ... */"
* Use Apple's markup format (https://developer.apple.com/library/content/documentation/Xcode/Reference/xcode_markup_formatting_ref/). In particular, surround code examples with three backticks.

(same comment applies to a few other comments that you have before declarations in this file)

marcrasi commented on Scripts/MNISTdataset.swift:
```diff
@@ -0,0 +1,97 @@
+let sh = Python.import("shutil")
+
+// MNIST constants
+let constants: [String: String] = [
```
> Declaring an enum without cases is the canonical way to define a “namespace” to group a set of related declarations, such as constants or helper functions.

(swift style guide)

An advantage over this `[String: String]` approach is that the compiler will catch references to missing constants at compile time!

The enum should be `fileprivate`.

marcrasi commented on Scripts/MNISTdataset.swift:
```diff
@@ -0,0 +1,97 @@
+import Python
+import TensorFlow
+
+let np = Python.import("numpy")
```
I think that `fileprivate` would be a good access level for all of these. If this code ever gets put in a module, these are probably not things that we want to expose to other files in the module.

Though I echo @saeta's point that it would be nice to do all of these using swift packages. You could probably add some dependencies to Package.swift.

marcrasi commented on Scripts/MNISTdataset.swift:
```diff
@@ -0,0 +1,97 @@
+]
+
+// Function to download and extract files from online
+func downloadAndExtract(){
```
If you make `readMNIST` call this automatically, then this can be `fileprivate`.

Put a space between the `)` and `{`.

marcrasi commented on Scripts/MNISTdataset.swift:
```diff
@@ -0,0 +1,97 @@
+
+// Function to download and extract files from online
+func downloadAndExtract(){
+    
```
We don't typically put blank lines at the beginning/end of code blocks. Only use blank lines to group statements into logical sections (https://google.github.io/swift/#vertical-whitespace).

marcrasi commented on Scripts/MNISTdataset.swift:
```diff
@@ -0,0 +1,97 @@
+        } else {
+            print("File \(extractedFile): already exists")
+        }
+         counter = counter + 1
```
this has an extra space in front of it

marcrasi commented on Scripts/MNISTdataset.swift:
```diff
@@ -0,0 +1,97 @@
+}
+
+// Reads MNIST images and labels from specified file paths.
+func readMNIST(trainImages: String, trainLabels: String, 
```
A good line wrapping (https://google.github.io/swift/#line-wrapping) would be:
```
func readMNIST(
    trainImages: String,
    trainLabels: String,
    testImages: String,
    testLabels: String
) -> (
    trainImages: Tensor<Float>,
    trainLabels: Tensor<Int32>,
￼    testImages: Tensor<Float>,
    testLabels: Tensor<Int32>
) {
```

marcrasi commented on Scripts/MNISTdataset.swift:
```diff
@@ -0,0 +1,97 @@
+}
+
+// Reads a file into an array of bytes.
+func readFile(_ filename: String) -> [UInt8] {
```
This can also be `fileprivate`

marcrasi commented on Scripts/MNISTdataset.swift:
```diff
@@ -0,0 +1,97 @@
+// Function to download and extract files from online
+func downloadAndExtract(){
+    
+    let helperFiles = [constants["trainImagesFile"]!, constants["trainLabelsFile"]!,
```
I think these should be one per line, like in function call line wrapping: https://google.github.io/swift/#function-calls

marcrasi commented:
Also, I talked to @rxwei and we agreed that actually the best location for this would be a new directory `Sources/DeepLearningContrib`, because it's a new kind of feature that we haven't developed standard protocols for yet. Once standard protocols for data loaders get designed, data loaders in `Sources/DeepLearningContrib` can be updated to conform and then moved into `Sources/DeepLearning`.

Sorry for asking you to move around the file so many times!

pruthiviraj71 commented:
Hey @marcrasi , I made some changes as suggested by you and @saeta to implement the script in native swift so I removed all python implementation and replaced it by swift alternative code. All that is left is downloading and extracting of gz files which I believe will be much easier with third party swift libraries or if you have any solution, I'll try implementing that.
And thanks for the code review!

marcrasi commented on Sources/DeepLearningContrib/MNISTdataset.swift:
```diff
@@ -0,0 +1,86 @@
+    let path = FileManager.default.currentDirectoryPath
+    for helperFile in helperFiles {
+        let url = NSURL(fileURLWithPath: path)
+        if let pathComponent = url.appendingPathComponent(helperFile) {
```
With `URL`, this is non-optional, so it should not require an `if let`!

Also I think `pathComponent` should be named something like `helperFileLocalUrl`, to describe that it's the local system url where the file should be.

marcrasi commented on Sources/DeepLearningContrib/MNISTdataset.swift:
```diff
@@ -0,0 +1,86 @@
+            }
+        }    
+    }
+    for extractedFile in extractedFiles {
```
It might simplify the code to have a single loop that does downloading and extracting. With a single loop, you won't have to reconstruct intermediate variables like `helperFileLocalUrl` that you need in both steps.

marcrasi commented on Sources/DeepLearningContrib/MNISTdataset.swift:
```diff
@@ -0,0 +1,86 @@
+                /// TODO: Add Swift Implementation for extracting downloaded files.
+                print("Extraction Completed: \(extractedFile)")
+            }
+            counter = counter + 1
```
Doing `for (index, bla) in array.enumerated()` would be more canonical than manually incrementing a `counter`. Though if you're not using counter anywhere, you should just delete it.

marcrasi commented on Sources/DeepLearningContrib/MNISTdataset.swift:
```diff
@@ -0,0 +1,86 @@
+    let trainLabels = try! Data(contentsOf: URL(fileURLWithPath: constants.trainLabels)).dropFirst(8).map { Int32($0) }
+    let testImages = try! Data(contentsOf: URL(fileURLWithPath: constants.testImages)).dropFirst(16).map { Float($0) }
+    let testLabels = try! Data(contentsOf: URL(fileURLWithPath: constants.testLabels)).dropFirst(8).map { Int32($0) }
+    let trainrowCount = Int32(trainLabels.count)
```
`row` should be capitalized for camelCase (also on the next line)

pruthiviraj71 commented:
Hey @marcrasi , I made few modifications as you requested and added gzip extraction process using the framework you referred to. Please provide feedback!

pruthiviraj71 commented on Sources/DeepLearningContrib/MNISTdataset.swift:
```diff
@@ -0,0 +1,86 @@
+/// MNIST constants
+fileprivate enum constants {
+    static let URL = "http://yann.lecun.com/exdb/mnist/"
+    static let trainImagesFile = "train-images-idx3-ubyte.gz"
```
Also @marcrasi , I am not sure if struct is better here because enum without cases does the work here. If you want, I can update it to use struct.

saeta commented:
Hi @pruthiviraj71 ! Thank you so much for this PR. I think for now, we'll close out this PR, as we have an alternate MNIST dataset implementation available as part of [swift-models](https://github.com/tensorflow/swift-models). If you'd like to discuss more of the design, please do let me know and we can schedule some time at an upcoming Swift design meetings! -Brennan
